invoices/settlement_interceptor.go

```diff
@@ -0,0 +1,224 @@
+// invoice is submitted to this service. This session allows the caller to block
+// until the invoice is processed.
+type InterceptSession struct {
+	HtlcModifyRequest
+
+	// ClientResponseChannel is a channel that is populated with the
+	// client's interceptor response during an interceptor session.
+	ClientResponseChannel chan HtlcModifyResponse
+
+	// ClientErrChannel is a channel that is populated with any errors that
+	// occur during the client's interceptor session.
+	ClientErrChannel chan error
+
+	// Quit is a channel that is closed when the session is no longer
+	// needed.
+	Quit chan struct{}
+}
+
+// HtlcModificationInterceptor is a service that intercepts HTLCs that aim to
+// settle an invoice, enabling a subscribed client to modify certain aspects of
+// those HTLCs.
+type HtlcModificationInterceptor struct {
+	wg sync.WaitGroup
+
+	// mu is a mutex that protects the callback and activeSessions fields.
+	mu sync.Mutex
+
+	// callback is the client callback function that is called when an
+	// invoice is intercepted. This function gives the client the ability to
+	// determine how the invoice should be settled.
+	callback HtlcModifyCallback
+
+	// activeSessions is a map of active intercept sessions that are used to
+	// manage the client query/response for a given invoice payment hash.
+	activeSessions map[CircuitKey]InterceptSession
+}
+
+// NewHtlcModificationInterceptor creates a new HtlcModificationInterceptor.
+func NewHtlcModificationInterceptor() *HtlcModificationInterceptor {
+	return &HtlcModificationInterceptor{
+		activeSessions: make(map[CircuitKey]InterceptSession),
+	}
+}
+
+// Intercept generates a new intercept session for the given invoice HTLC. The
+// session is returned to the caller so that they can block until the client
+// resolution is received.
+func (s *HtlcModificationInterceptor) Intercept(
+	clientRequest HtlcModifyRequest) fn.Option[InterceptSession] {
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// If there is no client callback set we will not handle the invoice
+	// further.
+	if s.callback == nil {
+		return fn.None[InterceptSession]()
+	}
+
+	// Create and store a new intercept session for the invoice. We will use
+	// the payment hash as the storage/retrieval key for the session.
+	sessionKey := clientRequest.ExitHtlcCircuitKey
+	session := InterceptSession{
+		HtlcModifyRequest:     clientRequest,
+		ClientResponseChannel: make(chan HtlcModifyResponse, 1),
+		ClientErrChannel:      make(chan error, 1),
+		Quit:                  make(chan struct{}, 1),
+	}
+	s.activeSessions[sessionKey] = session
+
+	// The callback function will block at the client's discretion. We will
+	// therefore execute it in a separate goroutine.
+	s.wg.Add(1)
+	go func(cb HtlcModifyCallback) {
+		defer s.wg.Done()
+
+		// By this point, we've already checked that the client callback
+		// is set. However, if the client callback has been set to nil
+		// since that check then Exec will return an error.
```

Comment:
Was left over from a previous iteration where there was a safe callback. Removed it because I didn't understand it back then, but brought it back because it seems to be needed.